[wasm] Run `Wasm.Build.Tests` against workloads

radical opened the pull request:
Currently we run `Wasm.Build.Tests` on helix, by setting various paths to reference files from `artifacts`. This PR adds another mode, we *try* to test against workload packs.

For this:

1. we need the nugets for the various workload packs, and the manifest
2. manifest id, path to those nugets
3. and a SDK version to test

Steps:

1. The specified SDK is installed in `artifacts/bin/dotnet-workload`
2. the specified manifest is installed from the built nugets, and any `depends-on` manifests in the json.
3. `dotnet workload install` is used to install the packs, with the built nugets path being used as one the nuget sourcs

Once this setup is done, then the tests are run in an environment such that they use `dotnet` from above, and try to resolve packs from there.

- We still want to test without packs, the case of using `EMSDK_PATH` directly, for example, in the library tests. So, we now run `Wasm.Build.Tests` for Workloads, *and* the regular `EMSDK_PATH`.

- Two workload specific tests were added in a recent PR (https://github.com/dotnet/runtime/pull/54936)
- simple test for building blazorwasm template project with AOT
- a test to validate that the files in `UnixFilePermissions.xml` for the packs actually exist on disk

.. and these are enabled here.

- To use this, the project sets `TestUsingWorkloads=true`, and specifies the workload with:

```xml
    <WorkloadIdForTesting Include="wasm-tools"
                           Name="microsoft.net.workload.mono.toolchain"
                           ManifestName="Microsoft.NET.Workload.Mono.ToolChain"
                           Version="$(PackageVersion)"
                           VersionBand="$(SdkBandVersion)" />
```

Notes:
- This also fixes using the correct runtime pack based on the version in the manifest json.
- The sdk version is specified with `<SdkVersionForWorkloadTesting>6.0.100-preview.7.21326.4</SdkVersionForWorkloadTesting>` in `eng/Versions.props`
- Known issues: I'm explicitly invoking `Pack` on the aotcross pkgproj

TODO:
- Run the blazorwasm test with playwright
- Add sdk-with-no-packs case also

radical commented:
cc @lewing @steveisok RFC!

lewing commented on src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Sdk/Sdk/AutoImport.props.in:
```diff
@@ -0,0 +1,8 @@
+<Project>
+  <PropertyGroup>
+    <WasmNativeWorkload>true</WasmNativeWorkload>
+    <UseMonoRuntime>true</UseMonoRuntime>
```
But the right place to still all these is in Sdk.props I think

steveisok commented on src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in:
```diff
@@ -43,4 +48,21 @@
+            LatestRuntimeFrameworkVersion="${PackageVersion}"
+            RuntimePackRuntimeIdentifiers="linux-x64;osx-x64;browser-wasm"
+            RuntimePackNamePatterns="Microsoft.NETCore.App.Runtime.Mono.**RID**"
+            RuntimePackLabels="BrowserWorkload"
```
I also think the label might be important for the mono runtime packs. 

@akoeplinger would know for sure.

radical replied:
So, label should be `MonoWorkload`?

Also, is this correct/enough - `RuntimePackRuntimeIdentifiers="linux-x64;osx-x64;browser-wasm`?

lewing replied:
> I also think the label might be important for the mono runtime packs.
> 
> @akoeplinger would know for sure.

The label is definitely important,  the details are important here

steveisok replied:
I think we should override the existing mono `KnownRuntimePack` by first removing it and then replacing with a copy.  The only difference should be the `LatestRuntimeFrameworkVersion`, which should match what's in the workload.

The original:
https://github.com/dotnet/installer/blob/823ff33acf9c67374d4b1211850b22d3db7f020d/src/redist/targets/GenerateBundledVersions.targets#L314-L321

steveisok replied:
I also wonder if we need `<Target Name="UpdateFrameworkReferences" BeforeTargets="ProcessFrameworkReferences">`?  We may just be able to condition an ItemGroup and do all the magic there.

steveisok replied:
Perhaps something like:

```
<PropertyGroup Condition="'$(TargetPlatformIdentifier)' == 'android' or '$(TargetPlatformIdentifier)' == 'ios' or '$(TargetPlatformIdentifier)' == 'maccatalyst'
                            or '$(TargetPlatformIdentifier)' == 'tvos' or ($(RuntimeIdentifier)' == 'browser-wasm' and '$(UsingBrowserRuntimeWorkload)' == 'true')">
    <_MonoWorkloadTargetsMobile>true</_MonoWorkloadTargetsMobile>
    <_MonoWorkloadRuntimePackPackageVersion>${PackageVersion}</_MonoWorkloadRuntimePackPackageVersion>
  </PropertyGroup>
  
  <ItemGroup Condition="'$(_MonoWorkloadTargetsMobile)' == 'true'">
    <KnownRuntimePack Remove="Microsoft.NETCore.App" />
    <KnownRuntimePack Include="Microsoft.NETCore.App"
                      TargetFramework="net6.0"
                      RuntimeFrameworkName="Microsoft.NETCore.App"
                      LatestRuntimeFrameworkVersion="$(_MonoWorkloadRuntimePackPackageVersion)"
                      RuntimePackNamePatterns="Microsoft.NETCore.App.Runtime.Mono.**RID**"
                      RuntimePackRuntimeIdentifiers="@(MonoRuntimePackRids, '%3B')"
                      RuntimePackLabels="Mono"
                      />
  </ItemGroup>
  ```

radical replied:
That seems to work fine

radical commented on src/mono/nuget/Microsoft.NET.Runtime.WebAssembly.Sdk/Sdk/AutoImport.props.in:
```diff
@@ -0,0 +1,8 @@
+  <PropertyGroup>
+    <WasmNativeWorkload>true</WasmNativeWorkload>
+    <UseMonoRuntime>true</UseMonoRuntime>
+    <RuntimePackInWorkloadVersion>${PackageVersion}</RuntimePackInWorkloadVersion>
```
How do I access this version during the build?

lewing replied:
which build?

radical replied:
A user project build, or tests. I need the version for the tests, and right now it is being used to figure out the path for the runtime pack too. Though as you had suggested, we could move to using `%(ResolvedRuntimePack.PackageDirectory)`, but that needs some changes.

radical commented:
Just doing some final tidbits, but this should be good for reviewing.

steveisok commented on src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/Sdk.props:
```diff
@@ -1,4 +1,8 @@
 <Project>
+  <PropertyGroup>
```
As I understand, coreclr is going to be the default target for xamarin macos.  So, blanket setting this property is not going to work.  

There is the ability to switch runtimes, but I'm not sure if that is going to be a supported feature or not.

radical replied:
Hmm so, RID=browser-wasm for now?

radical commented:
/azp run

radical commented:
/azp run runtime,runtime-staging

radical committed:
[wasm] Emcc.props -> emcc-props.json

.. and use that instead of importing the props file. This allows reading
the file on demand, and can be delayed till we have the runtime pack
path.

radical committed:
[wasm] Build samples, functional tests, and Wasm.Build.Tests only when

.. needed.

- Don't build samples, functional tests, and Wasm.Build.Tests on EAT, AOT lanes
- And build the samples in parallel

radical committed:
[wasm] Use %(ResolvedRuntimePack.PackageDirectory) to get the runtime

.. pack, instead of constructing it manually from nuget package path.

radical committed:
[wasm] Wasm.Build.Tests cleanup

.. and add per-test timeout.

radical committed:
cross-aot: Allow the project to be built directly for packing

- Currently, the cross aot packages get built on CI for the various
  configs, but we do get the binary built in artifacts.
- These changes allow a package to be generated based on that, and it
  will be useful for installing workloads that use such a package.

radical committed:
Add support for building Wasm.Build.Tests with workloads

Currently we run `Wasm.Build.Tests` on helix, by setting various paths to
reference files from `artifacts`. This PR adds another mode, we *try* to test
against workload packs.

For this:

1. we need the nugets for the various workload packs, and the manifests
2. manifest id, path to those nugets
3. and a SDK version to use for testing

Steps:

1. The specified SDK is installed in `artifacts/bin/dotnet-workload`
2. the specified manifest is installed from the built nugets, and any
   `depends-on` manifests in the json.

3. `dotnet workload install` is used to install the packs, with the built
   nugets path being used as one the nuget sources

Once this setup is done, the tests are run in an environment such that
they use `dotnet` from above, and try to resolve packs from there.

Known issues:
1. `Pack` target is explicitly invoked for the aotcross projects

We still want to test without packs, the case of using `EMSDK_PATH` directly,
for example, in the library tests. So, we now run `Wasm.Build.Tests` for
Workloads, *and* the regular `EMSDK_PATH`.

Two workload specific tests were added in a recent PR (https://github.com/dotnet/runtime/pull/54936)
- simple test for building blazorwasm template project with AOT
- a test to validate that the files in `UnixFilePermissions.xml` for the packs actually exist on disk

.. and these are enabled here.

Notes:
- This also fixes using the correct runtime pack based on the version in the manifest json.
- The sdk version is specified with
  `<SdkVersionForWorkloadTesting>6.0.100-preview.7.21326.4</SdkVersionForWorkloadTesting>`
  in `eng/Versions.props`

TODO:
- Run the blazorwasm test with playwright
- Add sdk-with-no-packs case also

radical committed:
[wasm] Add support for running Wasm.Build.Tests on windows

radical committed:
cleanup

radical committed:
Set PYTHONPATH for emsdk case

radical committed:
Testing the build

radical committed:
Add missing EnvironmentVariables.cs

radical committed:
Add doc for workloads testing

radical committed:
cleanup

radical committed:
Log directories in the dotnet sdk-manifests, and packs dir if installation fails

radical committed:
Fix typo

radical committed:
fix typo

radical committed:
Fix skipping tests on windows

radical committed:
Add a prefix to helix jobs for wasm.build.tests

radical committed:
re-enable all the jobs

radical committed:
cleanup

akoeplinger commented on src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/Sdk.props:
```diff
@@ -1,4 +1,8 @@
 <Project>
+  <PropertyGroup>
+    <UseMonoRuntime Condition="'$(RuntimeIdentifier)' == 'Browser-wasm'">true</UseMonoRuntime>
```
I think/hope this is no longer needed after my #55258

radical replied:
Your PR changes the InTree, and LocalBuild targets, which don't get used for the workloads.

radical replied:
Follow up PR? :) And if we are moving it to wasm sdk, then might as well move it to wasm props, then it won't be needed in InTree/LocalBuild either.

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-workloads-staging2

radical committed:
lock access to the stringbuilder for output, to prevent it from getting corrupted by concurrent access

radical committed:
Add a separate build job for wasm.build.tests

radical committed:
dummy commit in mono, to trigger the builds

radical commented:
`Windows x86` failure is https://github.com/dotnet/runtime/issues/55611

lewing committed:
Lock around ToString to ensure the handlers have flushed

lewing reviewed (approved):
Discussed some follow up with @radical but since we're already have a pending pr building off this the changes can happen there
